### ISSUE TITLE
Add Graphical User Interface

### ISSUE THREAD

**Issue by YasmineBOU** [sukeesh/Jarvis#421]
As students, we would like to contribute to your project.
Could we have some suggestions about the contributions that we can do?

**Comment by pnhofmann** [comment]
Of course ;).

* If you know/use windows, I'd be especially pleased, if you could improve our windows port ;).
* If you know something about natural language processing or want to learn something about it: Language processing can definitely be improved in Jarvis!
* Or you could improve Voice control (of course not writing real speech recognition; we simply use APIs like Google Cloud Speech).
* Or add a new feature. Anything you can think of that might be useful. Or improve an existing feature.

**Comment by YasmineBOU** [comment]
We would like to make you a suggestion if you agree and if you think that it can be useful.
So we suggest to make a graphic interface for Jarvis.

**Comment by pnhofmann** [comment]
Don't really know how a graphical interface for Jarvis should look like - but if you do: Sure; go for it. why not?

**Comment by sanketdhole** [comment]
I could help you for cross platform interface using kivy. But its not so much attractive. It can only be used for simplified interface.

**Comment by kenzaal** [comment]
Hello [sanketdhole](/sanketdhole) ,
Thank you for your proposal.
We know that is amateurism but we are thinking of using  Pygame for this interface.
Do you think that is an adequate choice ?

**Comment by Zml99** [comment]
Hi @kenzaal,
I'm new here and I don't know a lot of programing (I'm learning),  but I think using pygame it's actually a good idea.

**Comment by sanketdhole** [comment]
@kenzaal that might be better. But I didn't see lot of applications using pygames. I have just heard of it. So, really sorry that I won't be any helpful on that path. Below is one of my example of application using kivy.
![screenshot 11](https://user-images.githubusercontent.com/26873435/53302420-e7a91580-3883-11e9-9b75-7dc5bfe3a035.png)

**Comment by Zml99** [comment]
What you guys think if Jarvis were a zebra??

**Comment by sanketdhole** [comment]
@Zml99 What do you mean by that?

**Comment by Zml99** [comment]
Like Cortana is a blue circle, google assistant is 3 dots, siri is a mic, etc. I mean the icon of the voice assistant. So Jarvis could be a white and blue zebra, to make it a little bit more natural... @sanketdhole

**Comment by YasmineBOU** [comment]
Hi!

We have already started making the interface that looks like for now like shown in the screenshots.
The next step is to put in the bottom an entry field for the user and under that a small box where to put the answers.

We would like to know your opinions on that. 
And if you have other suggestions, let us know!

For the opening:

![jarvis_opening_image](https://user-images.githubusercontent.com/33510200/53404382-d829e980-39b5-11e9-9b61-0bcdb556d1ea.png)

Main window: (working on it)
![jarvis_main_window](https://user-images.githubusercontent.com/33510200/53404373-d52ef900-39b5-11e9-9a2d-64e25956d70c.png)

**Comment by kenzaal** [comment]
Hi,

We have tried to merge our code with the Jarvis source code to test if it works on it. However when we execute it with the command `./jarvis`, we get this message error:


```
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/kenza/Documents/Jarvis_bug/Jarvis-code-source/jarviscli/__main__.py", line 2, in <module>
    import Jarvis
  File "/home/kenza/Documents/Jarvis_bug/Jarvis-code-source/jarviscli/Jarvis.py", line 5, in <module>
    from windowClass import JarvisWindow
  File "/home/kenza/Documents/Jarvis_bug/Jarvis-code-source/jarviscli/windowClass.py", line 2, in <module>
    import pygame

```
**PS:**
 We used to execute our code with python3 and pygame 1.9.1release+dfsg-10 and it works with no problem.

**Comment by pnhofmann** [comment]
Do you know virtualenv?

https://docs.python-guide.org/dev/virtualenvs/

If you have multiple projects you don't want to install all python packages globally. So you commonly create a "virtualenv". You activate this virtualenv with: ``source env/bin/activate`` (in jarvis directory) and deactivate with ``deactivate``.
Think of it as a "namespace" - you get a clean python environment where you can install packages in any version without affecting anything else. 

So basically you have to add "pygame" to requirements.txt. Activate virtualenv ``source env/bin/activate``. Then install requirements with ``pip install -U -r requirements.txt``.

**Comment by JonanOribe** [comment]
> Hi!
> 
> We have already started making the interface that looks like for now like shown in the screenshots.
> The next step is to put in the bottom an entry field for the user and under that a small box where to put the answers.
> 
> We would like to know your opinions on that.
> And if you have other suggestions, let us know!
> 
> For the opening:
> 
> ![jarvis_opening_image](https://user-images.githubusercontent.com/33510200/53404382-d829e980-39b5-11e9-9b61-0bcdb556d1ea.png)
> 
> Main window: (working on it)
> ![jarvis_main_window](https://user-images.githubusercontent.com/33510200/53404373-d52ef900-39b5-11e9-9a2d-64e25956d70c.png)

Looks amazing!!

**Comment by YasmineBOU** [comment]
Hi!

We have tried to pull a version of the GUI but when the tests have been done, it finds that it doesn't match any of the python's versions. Is it because we have used Pygame?

**Comment by shreyashag** [comment]
Hi!

I'm new to this discussion, but just a question. 
How will the GUI work in case of docker deployment?

Wouldn't that need port forwarding the X server or something along the same lines? Also, how would this be achieved in Windows?

**Comment by shreyashag** [comment]
Makes sense! 
Is there any update on this @YasmineBOU ?

**Comment by YasmineBOU** [comment]
Hi!

As we are doing an internship and other stuffs, we haven't much time to continue working on it. However, after this internship finished and of course if everything goes well, we'll try to continue on it.

Please excuse us for the delay.

**Comment by pnhofmann** [comment]
No need for an excuse and please don't trouble yourself ;).

**Comment by YasmineBOU** [comment]
Thank you for your understanding.

**Comment by AhmetFurkann** [comment]
Hello everyone, does the graphical user interface development continue? I have looked at the branches, however I could not see a branch which is related to GUI.

I have an idea for the Jarvis GUI which is based on `PyQt5` and `PyOpenGL`. If it is allowed, can I implement the a graphical user interface for Jarvis? And for which branch should I open the pull request? To the master branch or to another branch?

**Comment by pnhofmann** [comment]
Hi @AhmetFurkann !

Here's my early test - but yeah, nothing special to see ;).

https://github.com/sukeesh/Jarvis/tree/dev/jarviscli/frontend/gui

(Dev-branch will hopefully merged into master the coming weeks)

I decided to use kivy mainly because it works really great on Android (and ios) and I thought Jarvis on my phone could be quite nice ;). 

> I have an idea for the Jarvis GUI which is based on PyQt5 and PyOpenGL.

But yeah, dev branch does even have a simple abstraction to run multiple front-ends. So if you like, I really wouldn't have any problems if you implement your own idea and would merge for sure ;). Just use dev-branch.

**Comment by AhmetFurkann** [comment]
First of all, I really thank your for your feedback @pnhofmann !

> Here's my early test - but yeah, nothing special to see ;).

In my perspective almost every code is worth to preview :).

All right, I checked and saw the frontend development section. I will try to implement my idea on the branch of dev. Probably you can see the development couple days later.

> I decided to use kivy mainly because it works really great on Android (and ios) and I thought Jarvis on my phone could be quite nice ;).

I agree with you, this is a really good idea :), on the other hand as I researched a little bit and found some ways to deploy a PyQt application to Android platforms and as well as other platforms like Windows and IOS. (https://stackoverflow.com/a/48172171)
